core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -704,7 +704,15 @@ public PlanWithProperties visitFilter(FilterNode node, PreferredProperties prefe
                 }
             }
 
-            return rebaseAndDeriveProperties(node, planChild(node, preferredProperties));
+            PlanWithProperties child = planChild(node, preferredProperties);
+
+            return rebaseAndDeriveProperties(node, new PlanWithProperties(
+                    child.getNode(),
+                    ActualProperties
+                            .builderFrom(child.getProperties())
+                            // Exclude order sensitive properties
+                            .local(child.getProperties().getLocalProperties().stream().filter(p -> !p.isOrderSensitive()).toList())
```

Comment:
Filter should not exclude any properties. I think the problem is due to an local exchange that's being added after WindowNode, which messes up the ordering guarantees on a global level, even if it preserves the ordering within the stream:

```
 Fragment 0 [SINGLE]
     Output layout: [quantity, row_number_0]
     Output partitioning: SINGLE []
     Output[columnNames = [quantity, row_number]]
     │   Layout: [quantity:double, row_number_0:bigint]
     │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
     │   row_number := row_number_0
     └─ Filter[filterPredicate = (mod("row_number_0", BIGINT '1000000') = BIGINT '0')]
        │   Layout: [quantity:double, row_number_0:bigint]
        │   Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
        └─ LocalExchange[partitioning = ROUND_ROBIN]         <<<<<<<<<<<<<<<
           │   Layout: [quantity:double, row_number_0:bigint]
           │   Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
           └─ Window[orderBy = [quantity ASC NULLS LAST]]
              │   Layout: [quantity:double, row_number_0:bigint]
              │   row_number_0 := row_number() RANGE UNBOUNDED_PRECEDING CURRENT_ROW
              └─ LocalExchange[partitioning = SINGLE]
                 │   Layout: [quantity:double]
                 │   Estimates: {rows: 6001215 (51.51MB), cpu: 0, memory: 0B, network: 0B}
                 └─ RemoteSource[sourceFragmentIds = [1]]
                        Layout: [quantity:double]
```

For example, here's another query that has the same issue and does not involve any filters:

```
SELECT row_number * 10 FROM (
    SELECT quantity, row_number() OVER (ORDER BY quantity) as row_number
    FROM tpch.sf1.lineitem
)
ORDER BY quantity;
```

```
 Fragment 0 [SINGLE]
     Output layout: [expr]
     Output partitioning: SINGLE []
     Output[columnNames = [_col0]]
     │   Layout: [expr:bigint]
     │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
     │   _col0 := expr
     └─ Project[]
        │   Layout: [expr:bigint]
        │   Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
        │   expr := ("row_number_0" * BIGINT '10')
        └─ LocalExchange[partitioning = ROUND_ROBIN]       <<<<<<<<<<<<<<<
           │   Layout: [quantity:double, row_number_0:bigint]
           │   Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
           └─ Window[orderBy = [quantity ASC NULLS LAST]]
              │   Layout: [quantity:double, row_number_0:bigint]
              │   row_number_0 := row_number() RANGE UNBOUNDED_PRECEDING CURRENT_ROW
              └─ LocalExchange[partitioning = SINGLE]
                 │   Layout: [quantity:double]
                 │   Estimates: {rows: 6001215 (51.51MB), cpu: 0, memory: 0B, network: 0B}
                 └─ RemoteSource[sourceFragmentIds = [1]]
                        Layout: [quantity:double]
```

I didn't look too closely, but I think it may be being caused by AddExchanges.visitSort removing the SortNode, and then AddLocalExchanges (incorrectly) adding a local exchange after the WindowNode.

Reply:
Thank you. I didn't understand the issue properly. I will try to revise the fix.

Reply:
In this case,

```sql
SELECT * FROM (
  SELECT l_quantity, row_number() OVER (ORDER BY l_quantity) as row_number 
  FROM tpch.sf1.lineitem
) WHERE mod(row_number, 1000000) = 0 ORDER BY l_quantity;
```

`FilterNode` requires unpartitioned parallel streams so `LocalExchange` is added here:

https://github.com/trinodb/trino/blob/91ee252b8eba1864e35661d57a854be06b98fe27/core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddLocalExchanges.java#L1047-L1057

In this case,

```sql
SELECT row_number * 10 FROM (
    SELECT quantity, row_number() OVER (ORDER BY quantity) as row_number
    FROM tpch.sf1.lineitem
)
ORDER BY quantity;
```

`AddExchanges.visitProject()` adds `LocalExchange` here:

https://github.com/trinodb/trino/blob/91ee252b8eba1864e35661d57a854be06b98fe27/core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddLocalExchanges.java#L232-L235

I wonder if there are other cases and how we can fix all cases.

Reply:
Take a look at AddExchanges.visitSort(). It's doing the following:

```java
 if (child.getProperties().isSingleNode()) {
     // current plan so far is single node, so local properties are effectively global properties
     // skip the SortNode if the local properties guarantee ordering on Sort keys
     // TODO: This should be extracted as a separate optimizer once the planner is able to reason about the ordering of each operator
     List<LocalProperty<Symbol>> desiredProperties = node.getOrderingScheme().toLocalProperties();

     if (LocalProperties.match(child.getProperties().getLocalProperties(), desiredProperties).stream()
             .noneMatch(Optional::isPresent)) {
         return child;
     }
 }
```

This bit of code is making the assumption that it's safe to remove the Sort because there are no exchanges *yet*. However, it's forgetting the fact that the data *needs* to be sorted, so later when the exchange is added this expectation is broken. 

I think the simplest solution is to get rid of that block of code altogether, as it's based on incorrect assumptions -- that the plan will continue to be "single node" after the SortNode is removed.

Reply:
I see. And this optimization needs to be applied after all exchanges are added if we would keep it.

Reply:
Yes, either that or the plan needs to retain information about required properties, but that’s a much larger and invasive change.

Reply:
Yeah, need to do almost same thing as `AddExchanges` but only for eliminating duplicated sort...

For now, I just removed sort elimination from `AddExchanges`.